### PR TITLE
[FIX] account: properly compute residual amounts during onchange

### PR DESCRIPTION
**Pull request by tdecaluwe**
The `_compute_amount_residual` method computes residual amounts on reconcilable account move lines. As it uses an sql query for an efficient computation it discerns stored records from new records (having a new id) by filtering out records with a falsy id field (new ids are always falsy). It then proceeds the computation for stored records only.

The bug arises when the compute method is called on records during an onchange call, as the records get reassigned a new id as well. This can be seen when opening the total due followup view and toggling the blocked field on any of the unreconciled entries. The problem is resolved by including records during an onchange in the computation: in practice the computation can simply be carried out on all `_origin` records.

opw-3388389

### CODE REVIEW

**Review by gawa-odoo — COMMENTED** [review]
Hello and thanks for the fix!
I think writing a test would be beneficial. It feels like something that could be overlooked in the future.

I will make a task for the banner that seems not logical either (with the 'With Overdue Invoices', if the partner has a single overdue invoice, if you block it, the banner stays up), as this ticket didn't really mention it (except if you want to try to fix it :smile: ).
I will also check later if we could have the same problem with the payment state on moves that seems to follow the same logic

**Comment by tdecaluwe** [comment]
@gawa-odoo The test is in enterprise https://github.com/odoo/enterprise/pull/45418

**Review by h4818 — COMMENTED** [review]
Thanks for the fix.
Please add/modify comments to guide developers on why `_origin` is used.

File: addons/account/models/account_move_line.py [hunk 1]
```diff
@@ -697,13 +697,12 @@ def _compute_amount_residual(self):
             for unreconciled lines, and something in-between for partially reconciled lines.
         """
         need_residual_lines = self.filtered(lambda x: x.account_id.reconcile or x.account_id.account_type in ('asset_cash', 'liability_credit_card'))
-        stored_lines = need_residual_lines.filtered('id')
```
[the remark this answers is not in the document]
Comment (tdecaluwe) [reply addons/account/models/account_move_line.py:700]: Yes we could, but seems to be a bit of a non-op now.

[the remark this answers is not in the document]
Comment (h4818) [reply addons/account/models/account_move_line.py:700]: it reduces the diff
```py
        # Run the residual amount computation on all lines stored in the db. By
        # using _origin, new records (with a NewId) are excluded and the
        # computation works automagically for virtual onchange records as well.
        stored_lines = need_residual_lines._origin
```

[the remark this answers is not in the document]
Comment (tdecaluwe) [reply addons/account/models/account_move_line.py:700]: Yes, one line more diff for removing one line of boilerplate seems like a good trade off to me. Anyways updated the PR...

**Comment by h4818** [comment]
@robodoo r+

**Pull request closed without merging**